runtime/near-vm-logic/src/logic.rs

```diff
@@ -413,6 +419,23 @@ impl<'a> VMLogic<'a> {
         Ok(())
     }
 
+    /// Reads public key from register (if length is `u64::MAX`) or memory
+    /// (otherwise).
+    ///
+    /// Returns `Err(_)` on errors related to reading the raw data; `Ok(Err(_))`
+    /// if the data could not be interpreted as a public key; and `Ok(Ok(_))` on
+    /// success.  The nested result is needed to maintain behaviour regarding
+    /// when different errors are reported.
```

Comment:
This is a footgun I feel. I wonder what was the reasoning behind decoding the `PublicKey` here, rather than keeping the original approach and changing the argument being passed from `Vec<u8>` to `&[u8]`?

Reply:
That said, I think the current approach is overall better if we can drop the double `Result`. I wonder if all we need here is a `TODO` basically to make this change at some future time when protocol versioning is less of a menial chore.

Reply:
I agree that this is a footgun.

OK, so here’s `promise_batch_action_stake` if we just change argument
in receipt_manager methods::

```rust
let public_key = get_memory_or_register!(self, public_key_ptr, public_key_len);

let (receipt_idx, sir) = self.promise_idx_to_receipt_idx_with_sir(promise_idx)?;
self.pay_action_base(ActionCosts::stake, sir)?;
self.receipt_manager.append_action_stake(receipt_idx, amount, &public_key)?;
```

Notice that a) public_key holds shared reference on self.registers
and b) pay_action_base takes exclusive reference on self.  This does
not compile.

A less footgun alternative is to just pay the price and allocate
a vector after getting the data.